paramak/shape.py

```diff
@@ -0,0 +1,1210 @@
+            "Z", etc. A custom axis can be set by setting a list of two XYZ
+            floats. Defaults to None.
+        tet_mesh (str, optional): If not None, a tet mesh flag will be added to
+            the neutronics description output. Defaults to None.
+        physical_groups (dict, optional): contains information on physical
+            groups (volumes and surfaces). Defaults to None.
+        cut (paramak.shape or list, optional): If set, the current solid will
+            be cut with the provided solid or iterable in cut. Defaults to
+            None.
+        intersect (paramak.shape or list, optional): If set, the current solid
+            will be interested with the provided solid or iterable of solids.
+            Defaults to None.
+        union (paramak.shape or list, optional): If set, the current solid
+            will be united with the provided solid or iterable of solids.
+            Defaults to None.
+    """
+
+    def __init__(
+        self,
+        points=None,
+        connection_type="mixed",
+        name=None,
+        color=(0.5, 0.5, 0.5),
+        material_tag=None,
+        stp_filename=None,
+        stl_filename=None,
+        azimuth_placement_angle=0.0,
+        workplane="XZ",
+        rotation_axis=None,
+        tet_mesh=None,
+        physical_groups=None,
+        cut=None,
+        intersect=None,
+        union=None,
+    ):
+
+        self.connection_type = connection_type
+        self.points = points
+        self.stp_filename = stp_filename
+        self.stl_filename = stl_filename
+        self.color = color
+        self.name = name
+
+        self.cut = cut
+        self.intersect = intersect
+        self.union = union
+
+        self.azimuth_placement_angle = azimuth_placement_angle
+        self.workplane = workplane
+        self.rotation_axis = rotation_axis
+
+        # neutronics specific properties
+        self.material_tag = material_tag
+        self.tet_mesh = tet_mesh
+
+        self.physical_groups = physical_groups
+
+        # properties calculated internally by the class
+        self.solid = None
+        self.render_mesh = None
+        # self.volume = None
+        self.hash_value = None
+        self.points_hash_value = None
+        self.x_min = None
+        self.x_max = None
+        self.z_min = None
+        self.z_max = None
+        self.graveyard_offset = None  # set by the make_graveyard method
+        self.patch = None
+
+    @property
+    def solid(self):
+        """The CadQuery solid of the 3d object. Returns a CadQuery workplane
+        or CadQuery Compound"""
+
+        ignored_keys = ["_solid", "_hash_value"]
+        if get_hash(self, ignored_keys) != self.hash_value:
+            self.create_solid()
+            self.hash_value = get_hash(self, ignored_keys)
+
+        return self._solid
+
+    @solid.setter
+    def solid(self, value):
+        self._solid = value
+
+    @property
```

Comment:
**BEST PRACTICE**

The `@property` and `@<property>.setter` decorators exist as a way to override the get and set function of a particular field of a class. For instance, it can be used to:

* lazy-load a value if it requires complex operations
* make a property read-only (don't implement its setter)

However, the usage here is unnecessary. Nothing is being done when querying the `cut` attribute and nothing is being done to `set` it. I'd recommend removing both of these functions and just having an instance attribute called `cut`. The same applies for other attributes in this file and in other files as well.

<p align="right">
    <a href="https://www.pullrequest.com/">
        <img width="15%" src="https://static.pullrequest.com/images/dashboard/pullrequest-logo.svg">
    </a>
</p>